projects/packages/forms/src/modules/file-field/view.js

```diff
@@ -51,116 +113,166 @@ const addFileToContext = file => {
 	}
 
+	// Get all files that don't have an error properly
+	const validFiles = context.files.filter( fileInfo => ! fileInfo.error );
+
 	// Check if the user is trying to add more files then allowed.
-	if ( context.maxFiles < context.files.length + 1 ) {
+	if ( context.maxFiles < validFiles.length + 1 ) {
 		error = config.i18n.maxFiles;
 	}
 
-	const fileId = performance.now() + '-' + Math.random();
+	const clientFileId = performance.now() + '-' + Math.random();
 
 	// Update the context.
 	context.hasFiles = true;
+
 	context.files.push( {
 		name: file.name,
 		formattedSize: formatBytes( file.size, 2 ),
-		hasToken: false,
+		isUploaded: false,
 		hasError: !! error,
-		id: fileId,
+		id: clientFileId,
 		error,
 	} );
 
+	const uploadFileWithScope = withScope( uploadFile.bind( this, file, clientFileId ) );
+
 	// Start the upload if we don't have any errors.
-	! error && uploadFile( file, fileId );
+	! error && uploadFileWithScope();
 
 	// Load the file so we can display it. In case it is an image.
 	reader.onload = withScope( () => {
-		updateFileContext( { url: 'url(' + reader.result + ')' }, fileId );
+		updateFileContext( { url: 'url(' + reader.result + ')' }, clientFileId );
 	} );
 };
 
+// Map to store AbortControllers for each file upload
+const uploadControllers = new Map();
+
 /**
  * Make the endpoint request.
+ * This function is a generator so that we can use the withScope function.
+ * And the context gets passed to the onProgress and onReadyStateChange functions.
  *
- * @param {File}   file   - The file to upload.
- * @param {string} fileId - The file ID.
+ * @param {File}   file         - The file to upload.
+ * @param {string} clientFileId - The client file ID.
+ * @yield {Promise<string>} The upload token.
  */
-const uploadFile = ( file, fileId ) => {
-	const { endpoint, uploadToken } = getConfig( NAMESPACE );
+function* uploadFile( file, clientFileId ) {
+	const { endpoint, i18n } = getConfig( NAMESPACE );
+
+	const token = yield getUploadToken();
+
+	if ( ! token ) {
+		updateFileContext( { error: i18n.uploadFailed, hasError: true }, clientFileId );
+		return;
+	}
+
 	const xhr = new XMLHttpRequest();
 	const formData = new FormData();
 
+	// Create an AbortController for this upload
+	const abortController = new AbortController();
+	uploadControllers.set( clientFileId, abortController );
+
 	xhr.open( 'POST', endpoint, true );
-	xhr.upload.addEventListener( 'progress', withScope( onProgress.bind( this, fileId ) ) );
-	xhr.addEventListener( 'readystatechange', withScope( onReadyStateChange.bind( this, fileId ) ) );
+	xhr.upload.addEventListener( 'progress', withScope( onProgress.bind( this, clientFileId ) ) );
+	xhr.addEventListener(
+		'readystatechange',
+		withScope( onReadyStateChange.bind( this, clientFileId ) )
+	);
+
+	// Handle abort signal
+	abortController.signal.addEventListener( 'abort', () => {
+		xhr.abort();
+		updateFileContext( { error: i18n.uploadFailed, hasError: true }, clientFileId );
```

Comment:
Have you tested this? As `updateFileContext` accesses the context, and this function is executed outside of the original scope (asynchronously), it is likely that it needs `withScope`.

Reply:
Yes ;) 
We mostly wanted to have a way to cancel the network request (especially if it was a long one) 
I removed the `updateFileContext()` call since we can't display the errors. It might be neat to show them somehow but in a different iteration.